dropshot_endpoint/src/metadata.rs

```diff
@@ -239,6 +285,87 @@ impl ValidatedEndpointMetadata {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum VersionRange {
+    All,
+    From(semver::Version),
+    Until(semver::Version),
+    FromUntil(semver::Version, semver::Version),
+}
+
+fn parse_semver(v: &syn::LitStr) -> syn::Result<semver::Version> {
+    v.value()
+        .parse::<semver::Version>()
+        .map_err(|e| {
+            syn::Error::new_spanned(v, format!("expected semver: {}", e))
+        })
+        .and_then(|s| {
+            if s.pre == semver::Prerelease::EMPTY {
+                Ok(s)
+            } else {
+                Err(syn::Error::new_spanned(
+                    v,
+                    String::from(
+                        "semver pre-release string is not supported here",
+                    ),
+                ))
+            }
+        })
+        .and_then(|s| {
+            if s.build == semver::BuildMetadata::EMPTY {
+                Ok(s)
+            } else {
+                Err(syn::Error::new_spanned(
+                    v,
+                    String::from("semver build metadata is not supported here"),
+                ))
+            }
+        })
+}
+
+impl syn::parse::Parse for VersionRange {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(syn::Token![..]) {
+            let _ = input.parse::<syn::Token![..]>()?;
+            if input.is_empty() {
+                Ok(VersionRange::All)
+            } else {
+                let latest = input.parse::<syn::LitStr>()?;
+                let latest_semver = parse_semver(&latest)?;
+                Ok(VersionRange::Until(latest_semver))
+            }
+        } else {
+            let earliest = input.parse::<syn::LitStr>()?;
+            let earliest_semver = parse_semver(&earliest)?;
+            let _ = input.parse::<syn::Token![..]>()?;
+            let lookahead = input.lookahead1();
+            if lookahead.peek(syn::LitStr) {
+                let latest = input.parse::<syn::LitStr>()?;
+                let latest_semver = parse_semver(&latest)?;
+                if latest_semver < earliest_semver {
+                    let span: TokenStream = [
+                        TokenTree::from(earliest.token()),
+                        TokenTree::from(latest.token()),
+                    ]
+                    .into_iter()
+                    .collect();
+                    return Err(syn::Error::new_spanned(
+                        span,
+                        String::from(
+                            "semver range (from ... until ...) has the \
+                             endpoints out of order",
```

Comment:
from version is after until version?

Reply:
Right.  Are you suggesting changing the message?  I'm fine with that (although I think it's a little hard to parse, using "from" and "until" as adjectives, especially if you're not already in the headspace of these arguments being called the "from version" and the "until version").

Reply:
I found the suggestion that the numbers are out of order potentially misleading. e.g. if I have 1.2.3..1.2.2 and what I actually meant was 1.2.3..1.2.12